core/src/main/scala/chisel3/Module.scala

```diff
@@ -597,6 +601,36 @@ package experimental {
       }
     }
 
+    /** Returns a FIRRTL ModuleTarget that references this object, relative to an optional root.
+      *
+      * If `root` is defined, the target is a hierarchical path starting from `root`.
+      *
+      * If `root` is not defined, the target is a hierarchical path equivalent to `toAbsoluteTarget`.
+      *
+      * @note If `root` is defined, and has not finished elaboration, this must be called within `atModuleBodyEnd`.
+      * @note The BaseModule must be a descendant of `root`, if it is defined.
+      * @note This doesn't have special handling for Views.
+      */
+    final def toRelativeTarget(root: Option[BaseModule]): IsModule = {
+      (root, _parent) match {
+        // If root was defined, and we are it, return this.
+        case (Some(_), _) if root.get == this => getTarget
+        // If root was defined, and we are not there yet, recurse up.
+        case (Some(definedRoot), Some(parent)) => parent.toRelativeTarget(root).instOf(this.instanceName, name)
+        // If root was defined, and we are at a ViewParent, use its absolute target.
+        case (Some(definedRoot), None) if this == ViewParent => ViewParent.absoluteTarget
+        // If root was defined, and there is no parent, the root was not an ancestor.
+        case (Some(definedRoot), None) =>
+          throwException(s"Requested .toRelativeTarget relative to ${definedRoot.name}, but it is not an ancestor")
+        // If root was not defined, and there is a parent, recurse up.
+        case (None, Some(parent)) => parent.toRelativeTarget(root).instOf(this.instanceName, name)
+        // If root was not defined, and we are at a ViewParent, use its absolute target.
+        case (None, None) if this == ViewParent => ViewParent.absoluteTarget
+        // If root was not defined, and there is no parent, return this.
+        case (None, None) => getTarget
+      }
```

Comment:
I feel like we could simplify this logic a bit, several of the cases seem basically duplicated, especially lines 619 and 626 as well as 621 and 628.
```scala
if (root.contains(this)) getTarget
else if (this == ViewParent) ViewParent.absoluteTarget
else (root, this._parent) match {
  case (_, Some(parent)) =>  parent.toRelativeTarget(root).instOf(this.instanceName, name)
  case (Some(definedRoot), None) =>
    throwException(s"Requested .toRelativeTarget relative to ${definedRoot.name}, but it is not an ancestor")
  case (None, None) => getTarget
}
```
Maybe? I just sketched that up really fast so it may not be correct

Reply:
I'll leave it to you to decide how to express this in a way that is the most clear, approved for whatever you prefer.

Reply:
Your suggestion works great as-is, just applied my old comments to the new structure. Thanks for taking the time to think hard enough about this to make that suggestion.